### ISSUE TITLE
Make the JS more CSP friendly

### ISSUE THREAD

**Issue by udf2457** [videojs/video.js#7887]
### Description

At present, the js file is not particularly CSP friendly, you receive errors such as:

`Content Security Policy: The page's settings blocked the loading of a resource at inline ("default-src").`

This is despite having appropriate headers set.

I believe this is the same as an issue I have experienced before with FontAwesome.  In that the JS file attempts to make CSS DOM manipulation which violates CSP (unless you use the `unsafe` flag, which is ill-advised for obvious reasons).

TL;DR, FontAwesome resolved this by providing the option to:

1. Disable automatic CSS insertion.
2. Reference the external CSS file explicitly.

(see https://fontawesome.com/v6/docs/web/dig-deeper/security#contentHeader)

They achieve this by:

1. Adding the option to use a `data-auto-add-css="false"` flag which disables automatic CSS insertion (e.g. `<script src="https://example.com/fontawesome/v6.1.2/js/all.js" data-auto-add-css="false"></script>`)
2. Providing a separate CSS file which can be referenced alongside (e.g. `<link href="https://example.com/fontawesome/v6.1.2/css/svg-with-js.css" rel="stylesheet">`)
i.e. your final code would look like:
```
      <script src="https://example.com/fontawesome/v6.1.2/js/all.js" data-auto-add-css="false"></script>
      <link href="https://example.com/fontawesome/v6.1.2/css/svg-with-js.css" rel="stylesheet">
```

I suggest the videojs team consider giving users the similar option (i.e. publish extracted CSS and provide ability to disable JS css insertion)

I have had zero CSP errors with FontAwesome since implementing the above with their library.

### Steps to reproduce

See Main NOte

### What version of Video.js are you using?

7.20.2

### What browser(s) including version(s) does this occur with?

All

### What OS(es) and version(s) does this occur with?

All

**Comment by gkatsev** [comment]
We do have an option to turn off dynamic style elements by setting `window.VIDEOJS_NO_DYNAMIC_STYLE = true`. Looks like it's only documented on the [skins guide](https://videojs.com/guides/skins/#disabling-additional-).
The dynamic styles are there for better fluid layout and to resize the player based on the intrinsic size of the video. Without it, as long as you size the player properly, it shouldn't be a problem.

As for other CSP issues, it gets a bit more complicated because we need to do things like set blob urls on video elements.

**Comment by udf2457** [comment]
I will try that out and let you know shortly @gkatsev 

Also are you aware video.js is not `require-trusted-types-for 'script'` safe ? 
(https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/require-trusted-types-for)

> Uncaught TypeError: Failed to set the 'innerHTML' property on 'Element': This document requires 'TrustedHTML' assignment.

**Comment by gkatsev** [comment]
Yeah, there's still a few more uses of innerHTML. Mostly in https://github.com/videojs/video.js/blob/main/src/js/tracks/text-track-settings.js, I believe. No one had the chance to address that yet. Any help would be appreciated.

**Comment by udf2457** [comment]
Thanks @gkatsev , I'm currently chest-deep in web dev at $work, but maybe if I have time after that !

Also, re: `window.VIDEOJS_NO_DYNAMIC_STYLE = true` that removes some errors, but it doesn't prevent your embedded font from trying to load ...

> Content Security Policy: The page's settings blocked the loading of a resource at data:application/font-woff;charset=utf-8…
> downloadable font: font load failed (font-family: "VideoJS" style:normal weight:400 stretch:100 src index:0): content blocked source: data:application/font-woff;charset=utf-

**Comment by gkatsev** [comment]
The fonts should be in the external CSS file and not set via JS.

**Comment by BasvanH** [comment]
Any update on this issue perhaps? Is making the library CSP friendly high on the to-do list?

**Comment by vincentjames501** [comment]
Is there any other approach to fixing this?

`Refused to create a worker from 'blob:http://<webserver>/f675bf28-a6df-4b59-b032-23e6e02326c6' because it violates the following Content Security Policy directive: `

Setting the video component's `src` attribute to `blob:*` is a large security issue for many of us.

Here is something super simple to reproduce the issue:

- Create a new directory and cd into it
- Create the following file called `index.html` in that directory
```html
<head>
    <meta
  http-equiv="Content-Security-Policy"
  content="script-src 'self' https://vjs.zencdn.net; style-src https://vjs.zencdn.net; media-src https://d2zihajmogu5jn.cloudfront.net" />
    <link href="https://vjs.zencdn.net/8.17.2/video-js.css" rel="stylesheet" />
</head>
  
  <body>
    <video
      id="my-video"
      class="video-js"
      controls
      preload="auto"
      width="640"
      height="264"
      data-setup="{}">
      <source src="https://d2zihajmogu5jn.cloudfront.net/bipbop-advanced/bipbop_16x9_variant.m3u8" type="application/x-mpegURL" />
    </video>
  
    <script src="http://localhost:8000/init.js"></script>
    <script src="https://vjs.zencdn.net/8.17.2/video.min.js"></script>
  </body>
```
- Create the following file called `init.js` in that directory
```javascript
window.VIDEOJS_NO_DYNAMIC_STYLE = true;
```
- From w/in that directory, simply run something like `python3 -m http.server` and navigate to http://localhost:8000 and look at the console error

<img width="1720" alt="image" src="https://github.com/user-attachments/assets/b82ec991-c012-4bfd-8e44-bc2eebb88935">

**Comment by udf2457** [comment]
I've long given up on videojs because of the dev's unwillingless to do anything about reality.  The fact this issue is still open two years later and people (@vincentjames501 ) are still having issues says it all.

Other people, from YouTube downwards have no problems getting videos displayed in a browser without erroring out on CSP issues.

I'm unsubscribing from this thread given I've no longer any interest in wasting my time on VideoJS.

**Comment by Cozmoz** [comment]
I’m seeing this still on Video.js 8.23.4 while testing a staged CSP rollout.

Current enforced baseline CSP is working, but when adding a stricter report-only policy with:

```http
style-src-attr 'none'
```

VideoJS reports runtime inline style attribute writes:

```
{
  "effective_directive": "style-src-attr",
  "violated_directive": "style-src-attr",
  "blocked_uri": "/inline",
  "document_uri": "/wall",
  "source_file": "/libs/video.js/8.23.4/video.min.js",
  "line_number": 12,
  "column_number": 216137,
  "disposition": "report"
}
```

The same report appears on pages with VideoJS players, e.g. /wall and /events.

So the remaining CSP blocker is not just JS-inserted CSS, but runtime writes to inline style attributes. A CSP-friendly mode would ideally avoid inline style attributes for static player state/layout where classes or CSS variables could be used, or expose a supported way for apps to opt into class-based styling.

We are keeping full CSP report-only for now because enforcing style-src-attr 'none' would risk breaking VideoJS-heavy media flows.

That gives maintainers the exact directive, version, source location, and practical impact. The original issue already asks for CSP-friendlier JS/CSS handling, but this adds the newer `style-src-attr` angle, which is the thing blocking us now.
